[Humble] Add topics with zero message counts to the SQLiteStorage::get_metadata().

MichaelOrlov opened the pull request:
- Closes https://github.com/ros2/rosbag2/issues/1719
- Analog and manual backport of the #1725. 
Have to do manual backporting because the code base is very different on humble branch in the affected places.

morlov-apexai committed:
Add topics with zero messages count to the SQLiteStorage::get_metadata()

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

fujitatomoya reviewed (commented):
@MichaelOrlov i am trying to address for `rolling` with https://github.com/ros2/rosbag2/pull/1725.

can i borrow your code for testing? i am not sure if this works for `rolling`... any suggestions?

MichaelOrlov commented:
@fujitatomoya Sure, feel free to take any parts for rolling. The test should work for rolling as well; however, it will be better to parametrize it to run it for the `mcap` storage as well.

morlov-apexai committed:
Update get_metadata_include_topics_with_zero_messages test after rebase

- Parametrize get_metadata_include_topics_with_zero_messages test to run
for mcap and sqlite3 storage plugins.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@fujitatomoya I've updated `get_metadata_include_topics_with_zero_messages` test to run it for mcap as well.
Feel free to take it for rolling. It should fit for rolling as is now. I assume that no modification needed.

MichaelOrlov commented on rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:
```diff
@@ -463,13 +463,25 @@ void SqliteStorage::prepare_for_reading()
+  if (database_->field_exists("topics", "offered_qos_profiles")) {
+    auto statement = database_->prepare_statement(
+      "SELECT name, type, serialization_format, offered_qos_profiles FROM topics ORDER BY id;");
+    auto query_results =
+      statement->execute_query<std::string, std::string, std::string, std::string>();
 
-  for (auto result : query_results) {
-    all_topics_and_types_.push_back(
-      {std::get<0>(result), std::get<1>(result), std::get<2>(result), ""});
+    for (auto result : query_results) {
+      all_topics_and_types_.push_back(
+        {std::get<0>(result), std::get<1>(result), std::get<2>(result), std::get<3>(result)});
+    }
+  } else {
+    auto statement = database_->prepare_statement(
+      "SELECT name, type, serialization_format FROM topics ORDER BY id;");
+    auto query_results = statement->execute_query<std::string, std::string, std::string>();
+
+    for (auto result : query_results) {
+      all_topics_and_types_.push_back(
+        {std::get<0>(result), std::get<1>(result), std::get<2>(result), ""});
+    }
```
Note for reviewers. This if-else statement was missing on humble because we didn't backport https://github.com/ros2/rosbag2/pull/1722. However, it is required to correctly fill topics and types map with `offered_qos_profiles`.

fujitatomoya reviewed (approved):
lgtm with green CI.

@MichaelOrlov thanks for the note. btw, this is also gonna go to `iron` after this humble fix, right? because https://github.com/ros2/rosbag2/pull/1725 only addressed rolling and jazzy.

MichaelOrlov commented:
@fujitatomoya Thanks for the review. We can propagate either this PR to the Iron or #1725. Need to double-check the base sources on Iron to see which one is better fit.

MichaelOrlov commented:
Pulls: ros2/rosbag2#1722
Gist: https://gist.githubusercontent.com/MichaelOrlov/1a1a4f8ac0ad7670f8603b60e4eef957/raw/40ce70e63ea8ce5ccc15d405be9bd9194282d1d4/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_cpp rosbag2_storage_default_plugins rosbag2_tests
TEST args:  --packages-above rosbag2_cpp rosbag2_storage_default_plugins rosbag2_tests
ROS Distro: humble
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14288
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21537)](http://ci.ros2.org/job/ci_linux/21537/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15821)](http://ci.ros2.org/job/ci_linux-aarch64/15821/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1032)](http://ci.ros2.org/job/ci_linux-rhel/1032/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22254)](http://ci.ros2.org/job/ci_windows/22254/)

MichaelOrlov commented:
The RHEL9 build fails with error message
```
--- stderr: rosbag2_storage_default_plugins
14:30:34 [K/home/jenkins-agent/workspace/ci_linux-rhel/ws/src/ros2/rosbag2/rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.c[K In member functio[Kvirtual rosbag2_storage::BagMetadata rosbag2_storage_plugins::SqliteStorage::get_metadat[K’:
14:30:34 [K/home/jenkins-agent/workspace/ci_linux-rhel/ws/src/ros2/rosbag2/rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:539:[Kerro[Kfind[K’ is not a member o[K[K’; did you mea[Kf[K’?
14:30:34   539 |       auto it = std::[Kfind[K(
14:30:34       |                      [K^~~~[K
14:30:34       |                      [Kf[K
```
@clalancette Do you have any idea what RHEL complaining about?

morlov-apexai committed:
Attempt to fix RHEL9 build failure

- Include <algorithm> header and explicitly use reference on
topic_metadata in a lambda capture list.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Re-run RHEL CI job after attempt to fix a build error.
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1041)](http://ci.ros2.org/job/ci_linux-rhel/1041/)

MichaelOrlov commented:
@clalancette I've fixed the RHEL build. I would appreciate if you would approve this backporting PR additionally so I can merge it.

clalancette commented:
> @clalancette I've fixed the RHEL build. I would appreciate if you would approve this backporting PR additionally so I can merge it.

So unfortunately, that succeeding build on Humble is for RHEL-9, not RHEL-8 (which is what Humble supports).  I think you should re-run the job on RHEL-8, and see what happens.

MichaelOrlov commented:
Run CI for RHEL**8**
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1049)](http://ci.ros2.org/job/ci_linux-rhel/1049/)

morlov-apexai committed:
Suppress PEP8 warning about putting module level imports on top of file

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
The warning message 
```
Failed
rosbag2_py.flake8.E402 (./test/test_convert.py:30:1)

Failing for the past 1 build (Since [#1049](https://ci.ros2.org/job/ci_linux-rhel/1049/) )
Error Message
module level import not at top of file:
from rosbag2_py import (
```
is unrelated to the changes in this PR.
However, I've made a rebase and added suppression for those warning.

Re-run CI for RHEL**8** after the rebase and suppressing PEP8:E402 warning
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1050)](http://ci.ros2.org/job/ci_linux-rhel/1050/)

MichaelOrlov commented:
@clalancette Ok. Now the RHEL**8** CI job is green. Ready to be approved.

fujitatomoya commented:
either @clalancette or @MichaelOrlov can you create the backport PR by mergifyio? (i do not have permission on this repo)

clalancette commented:
> either @clalancette or @MichaelOrlov can you create the backport PR by mergifyio? (i do not have permission on this repo)

I'm not sure what you mean; this one was the backport PR, no?

fujitatomoya commented:
@clalancette sorry for the complication,

- rolling, https://github.com/ros2/rosbag2/pull/1725
- jazzy, https://github.com/ros2/rosbag2/pull/1731 (backport of rolling)
- iron, missing
- this PR is for humble

i have consulted with @MichaelOrlov before, because of the base code difference, we did take the different pathes to `rolling/jazzy` and `humble/iron`. so that plan was to develop `humble` with this one and backport to `iron` which is missing now.

MichaelOrlov commented:
@mergifyio backport iron
